### DISCUSSION TITLE
Ruby serverless runtime always seems to be encoding to UTF-8

### DISCUSSION THREAD

**Discussion by lukel97** [vercel/vercel#4886] [Help]
Hi, I'm trying to use the ruby Serverless runtime to send back some binary data (specifically audio data), but even when the webrick handler runs fine locally there seems to be some implicit conversion going on behind the scenes that tries to convert the data to UTF-8. This is the simplest example I could find that has this:

```ruby
Handler = Proc.new do |req, res|
  res.body = "\xFF".b
end
```

This gives the error
```
 WEBrick 1.6.0
 ruby 2.7.1 (2020-03-31) [x86_64-linux]
 WEBrick::HTTPServer#start: pid=8 port=3000
127.0.0.1 - - [19/Jul/2020:12:25:08 UTC] "GET /api/speech HTTP/1.1" 200 1
- -> /api/speech
Error raised from handler method{
  "errorMessage": "\"\\xFF\" from ASCII-8BIT to UTF-8",
  "errorType": "Function<Encoding::UndefinedConversionError>",
  "stackTrace": [
  ]
 going to shutdown ...
 WEBrick::HTTPServer#start done.
```

Even when setting the content-type to 'audio/mpeg3' or another binary format, it always seems to try and convert it and fails. I've tried using `force_encoding` as well as setting `Encoding::default_external`/`Encoding::default_internal`

**Comment by styfle** [comment]
Try setting the content-type header like the following:

```rb 
require 'cowsay'

Handler = Proc.new do |req, res|
  name = req.query['name'] || 'World'

  res.status = 200
  res['Content-Type'] = 'text/text; charset=utf-8'
  res.body = Cowsay.say("Hello #{name}", 'cow')
end
```

https://vercel.com/docs/runtimes#official-runtimes/ruby

**Reply by lukel97** [reply]
It's the opposite problem I'm having haha, where even when the content-type header is set to something that isn't text and not utf-8, the runtime seems to be converting it to utf-8 anyway (which throws an exception on data that isn't utf-8 compatible, i.e. just binary data)

**Reply by styfle** [reply]
I see, that is strange indeed.

We got a PR #4593 from @jaredcwhite who might be able to provide more info here.

**Reply by jaredcwhite** [reply]
Hmm, I'd have to run some tests…to the best of my knowledge, if there's no `charset` specifically set for the return content type, nothing is automatically coerced to UTF-8, so I'm not quite sure where the issue lies.

**Reply by jaredcwhite** [reply]
I'm thinking maybe the issue is that binary data must be encoded as base64 in order to preserve formatting in the lambda output…something similar was done for the Python runtime. Ruby runtime doesn't contain that logic. https://github.com/vercel/vercel/blob/c1cdddd974e18e575d0150f778367b9823cd8df6/packages/now-python/now_init.py#L77

**Reply by styfle** [reply]
That sounds right 👍 

Does one of you want to create a PR to fix it?

**Comment by Beyarz** [comment]
Any updates on this? Experiencing this issue right now.

Edit: Nvm solved it.

**Reply by Patrick-ring-motive** [reply]
How?

**Reply by Beyarz** [reply]
By forcing the encoding and adding the following snippet to the content-type.
 
`body = message.force_encoding('utf-8')`
`response['Content-Type'] = 'text/plain; charset=utf-8'`

**Comment by Patrick-ring-motive** [comment]
This is still an issue. Is the workaround to encode binary data differently? I’d like to see how others solved this.

**Reply by Patrick-ring-motive** [reply]
This seems like a terrible workaround but this is what I have going now. For binary data that is not UTF8 I stringify using `body.unpack('C*').inspect` and then `new Uint8Array((JSON.parse(body)))` in my cloudflare worker.
